lib/maths/CTimeSeriesModel.cc

```diff
@@ -701,68 +737,69 @@ CUnivariateTimeSeriesModel::addSamples(const CModelAddSamplesParams& params,
+        m_RecentResiduals.pop_front();
+    }
+    TFloatMeanAccumulator next;
+    for (std::size_t i = 0; i < samples_.size(); ++i) {
+        double weight{maths_t::countForUpdate(weights_[i])};
+        next.add(samples_[i] / scales_[i], weight);
+    }
+    m_RecentResiduals.push_back({averageTime, next});
+    TDouble1Vec mean;
+    maths_t::TDoubleWeightsAry1Vec meanWeight;
+    std::tie(mean, meanWeight) = this->residualMean();
+
+    // Update the residual models.
     m_ResidualModel->addSamples(samples_, weights_);
     m_ResidualModel->propagateForwardsByTime(params.propagationInterval());
+    if (mean.size() > 0 && m_ResidualMeanModel != nullptr) {
+        m_ResidualMeanModel->addSamples(mean, meanWeight);
+        m_ResidualMeanModel->propagateForwardsByTime(params.propagationInterval());
+    }
+
     if (m_AnomalyModel != nullptr) {
         m_AnomalyModel->propagateForwardsByTime(params.propagationInterval());
     }
 
-    double multiplier{1.0};
-    if (m_Controllers != nullptr) {
-        TDouble1VecVec errors[2];
-        errors[0].reserve(samples.size());
-        errors[1].reserve(samples.size());
-        for (auto sample : samples_) {
-            this->appendPredictionErrors(params.propagationInterval(), sample, errors);
-        }
-        {
-            CDecayRateController& controller{(*m_Controllers)[E_TrendControl]};
-            core_t::TTime time{static_cast<core_t::TTime>(CBasicStatistics::mean(averageTime))};
-            TDouble1Vec trendMean{m_TrendModel->meanValue(time)};
-            multiplier = controller.multiplier(
-                trendMean, errors[E_TrendControl], this->params().bucketLength(),
-                this->params().learnRate(), this->params().decayRate());
-            if (multiplier != 1.0) {
-                m_TrendModel->decayRate(multiplier * m_TrendModel->decayRate());
-                LOG_TRACE(<< "trend decay rate = " << m_TrendModel->decayRate());
-            }
-        }
-        {
-            CDecayRateController& controller{(*m_Controllers)[E_ResidualControl]};
-            TDouble1Vec residualMean{m_ResidualModel->marginalLikelihoodMean()};
-            multiplier = controller.multiplier(
-                residualMean, errors[E_ResidualControl], this->params().bucketLength(),
-                this->params().learnRate(), this->params().decayRate());
-            if (multiplier != 1.0) {
-                m_ResidualModel->decayRate(multiplier * m_ResidualModel->decayRate());
-                LOG_TRACE(<< "prior decay rate = " << m_ResidualModel->decayRate());
-            }
-        }
-    }
+    double multiplier{this->updateDecayRates(params, averageTime, samples_)};
 
     if (m_Correlations != nullptr) {
         m_Correlations->addSamples(m_Id, params, samples, multiplier);
     }
 
     if (randomSample) {
-        m_SlidingWindow.push_back({randomSample->first, randomSample->second});
+        m_RecentSamples.push_back({randomSample->first, randomSample->second});
```

Comment:
could be an emplace_back?

Reply:
This is a boost circular buffer which doesn't support emplace_back unfortunately.